pkg/algorithm/hived_algorithm.go

```diff
@@ -559,57 +501,105 @@ func (h *HivedAlgorithm) setBadCell(c *PhysicalCell) {
+		if h.vcDoomedBadCellNum[vcName][c][l] <= 0 {
+			h.vcDoomedBadCellNum[vcName][c][l] = 0
+			if prevDoomedBadNum > 0 {
 				klog.Infof("Cell type %v (chain %v level %v) now has sufficient healthy cells (healthy %v, bad %v) "+
-					"for allocating the total free cells of all the VCs (%v).",
-					h.cellTypes[chain][level], chain, level, h.totalLeftCellNum[chain][level]-h.badFreeCellNum[chain][level],
-					h.badFreeCellNum[chain][level], h.allVCFreeCellNum[chain][level])
+					"for allocating the free cells of the VC %v (%v).",
+					h.cellTypes[c][l], c, l,
+					h.totalLeftCellNum[c][l]-h.badFreeCellNum[c][l], h.badFreeCellNum[c][l],
+					vcName, vcFreeCellNum[c][l])
 			}
-			for vcName, vcFreeCellNum := range h.vcFreeCellNum {
-				if vcFreeCellNum[chain][level] >= h.totalLeftCellNum[chain][level]-h.badFreeCellNum[chain][level] {
-					if vcFreeCellNum[chain][level] == h.totalLeftCellNum[chain][level]-h.badFreeCellNum[chain][level] {
-						klog.Infof("Cell type %v (chain %v level %v) now has sufficient healthy cells (healthy %v, bad %v) "+
-							"for allocating the free cells of the VC %v (%v).",
-							h.cellTypes[chain][level], chain, level,
-							h.totalLeftCellNum[chain][level]-h.badFreeCellNum[chain][level], h.badFreeCellNum[chain][level],
-							vcName, vcFreeCellNum[chain][level])
+		} else {
+			klog.Warningf("Cell type %v (chain %v level %v) now has fewer healthy cells (healthy %v, bad %v) "+
+				"than the free cells of the VC %v (%v). Certain cells in the VC are doomed to be bad.",
+				h.cellTypes[c][l], c, l,
+				h.totalLeftCellNum[c][l]-h.badFreeCellNum[c][l], h.badFreeCellNum[c][l],
+				vcName, vcFreeCellNum[c][l])
+		}
+		if prevDoomedBadNum > h.vcDoomedBadCellNum[vcName][c][l] {
+			numToReduce := prevDoomedBadNum - h.vcDoomedBadCellNum[vcName][c][l]
+			n := int32(0)
+			for _, vc := range h.vcSchedulers[vcName].getNonReservedFreeCellList()[c][l] {
+				virtualCell := vc.(*VirtualCell)
+				if virtualCell.GetPhysicalCell() == nil && virtualCell.GetAPIStatus().CellHealthiness == api.CellBad {
+					virtualCell.GetAPIStatus().CellHealthiness = api.CellHealthy
+					n++
+					if n >= numToReduce {
+						break
 					}
-					for _, vc := range h.vcSchedulers[vcName].getNonReservedFreeCellList()[chain][level] {
-						virtualCell := vc.(*VirtualCell)
-						if virtualCell.GetPhysicalCell() == nil && virtualCell.GetAPIStatus().CellHealthiness == api.CellBad {
-							virtualCell.GetAPIStatus().CellHealthiness = api.CellHealthy
-							break
-						}
+				}
+			}
+		} else if prevDoomedBadNum < h.vcDoomedBadCellNum[vcName][c][l] {
+			numToIncrease := h.vcDoomedBadCellNum[vcName][c][l] - prevDoomedBadNum
+			n := int32(0)
+			for _, vc := range h.vcSchedulers[vcName].getNonReservedFreeCellList()[c][l] {
+				virtualCell := vc.(*VirtualCell)
+				if virtualCell.GetPhysicalCell() == nil && virtualCell.GetAPIStatus().CellHealthiness != api.CellBad {
+					virtualCell.GetAPIStatus().CellHealthiness = api.CellBad
+					n++
```

Comment:
Also set all its children as Bad?
BTW, remind to also revert all its children when Health.